pkg/networkmanager/wireguard.jsx

```diff
@@ -0,0 +1,285 @@
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useContext, useEffect, useState } from 'react';
+import cockpit from 'cockpit';
+import { Button } from '@patternfly/react-core/dist/esm/components/Button/index.js';
+import { ClipboardCopy } from '@patternfly/react-core/dist/esm/components/ClipboardCopy/index.js';
+import { EmptyState, EmptyStateBody } from '@patternfly/react-core/dist/esm/components/EmptyState/index.js';
+import { FormGroup, FormFieldGroup, FormFieldGroupHeader, FormHelperText } from '@patternfly/react-core/dist/esm/components/Form/index.js';
+import { Grid } from '@patternfly/react-core/dist/esm/layouts/Grid/index.js';
+import { HelperText, HelperTextItem } from '@patternfly/react-core/dist/esm/components/HelperText/index';
+import { InputGroup } from '@patternfly/react-core/dist/esm/components/InputGroup/index.js';
+import { MinusIcon } from '@patternfly/react-icons';
+import { TextInput } from '@patternfly/react-core/dist/esm/components/TextInput/index.js';
+
+import { Name, NetworkModal, dialogSave } from "./dialogs-common";
+import { ModelContext } from './model-context';
+import { useDialogs } from 'dialogs.jsx';
+
+import './wireguard.scss';
+
+const _ = cockpit.gettext;
+
+function addressesToString(addresses) {
+    return addresses.map(address => address[0] + "/" + address[1]).join(", ");
+}
+
+function stringToAddresses(str) {
+    return str.split(/[\s,]+/).map(strAddress => {
+        const parts = strAddress.split("/");
+        if (parts.length > 2) {
+            throw new Error(_("Addresses is not formatted correctly"));
+        }
+        const [ip, prefix] = parts;
+        const defaultPrefix = "32";
+        // Gateway usually conflicts with routing that NetworkManager configures for WireGuard interfaces
+        // So it should not be set in that case or more accurately set to 0 i.e. "0.0.0.0" in string format
+        const gateway = "0.0.0.0";
+        return [ip, prefix ?? defaultPrefix, gateway];
+    });
+}
+
+export function WireGuardDialog({ settings, connection, dev }) {
+    const Dialogs = useDialogs();
+    const idPrefix = "network-wireguard-settings";
+    const model = useContext(ModelContext);
+
+    const [iface, setIface] = useState(settings.connection.interface_name);
+    const [privateKey, setPrivateKey] = useState("");
+    const [publicKey, setPublicKey] = useState("");
+    const [listenPort, setListenPort] = useState(settings.wireguard.listen_port);
+    const [addresses, setAddresses] = useState(addressesToString(settings.ipv4.addresses));
+    const [dialogError, setDialogError] = useState("");
+    const [peers, setPeers] = useState(settings.wireguard.peers.map(peer => ({ ...peer, allowedIps: peer.allowedIps.join(",") })));
+
+    // Additional check for `wg` after install_dialog for non-packagekit and el8 environments
+    useEffect(() => {
+        async function checkWireguardPackage() {
```

Comment:
that is weird. You should be able to simply do `useEffect(async () => { ...})` without the nested function.

Reply:
AFAIK that's not possible. async functions returns a promise object and react expects a clean up function in useEffect if we ever return anything.

```js
useEffect(() => {
  // effect code
  return () => {
    // return type == function
    // using useEffect(async () => {}) means return type == Promise object
  }
})
```
Found an example in the docs here https://react.dev/reference/react/useEffect#fetching-data-with-effects (please scroll down to the "You can also rewrite using the async / await syntax, but you still need to provide a cleanup function:" section)

Calling an `async` function like this is a pretty good pattern imho. It was suggested by Dan Abramov once here https://github.com/facebook/react/issues/14326#issuecomment-441680293. I've been using it ever since.

In this particular case, as it should be the first thing that runs and only in unmount/mount, we could take it out of `useEffect` into a separate function, similar to https://github.com/cockpit-project/cockpit-machines/blob/main/src/app.jsx#L76 and call it in `dialogError` like `const [dialogError, setDialogError] = useState(checkWireguardPackage());`. Ah it's asynchronous. Oops, sorry it wouldn't work. (Ignore this last paragraph :smile:)

Reply:
I see, thanks. TIL!